### PR TITLE
Add a compiletest header for edition

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @nikomatsakis 

Are the `-Zunstable-options` options needed in these tests? It looks like they aren't. If not, I can remove them.

### CODE REVIEW

File: src/tools/compiletest/src/header.rs [hunk 1]
```diff
@@ -647,6 +654,13 @@ impl Config {
     fn parse_run_rustfix(&self, line: &str) -> bool {
         self.parse_name_directive(line, "run-rustfix")
     }
+
+    fn parse_edition(&self, line: &str) -> Option<String> {
+        match line {
+            ed @ "2015" | "2018" => Some(ed.to_owned()),
+            _ => None,
+        }
+    }
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/header.rs]: maybe we should just permit you to write anything here?

File: src/tools/compiletest/src/header.rs [hunk 2]
```diff
@@ -647,6 +654,13 @@ impl Config {
     fn parse_run_rustfix(&self, line: &str) -> bool {
         self.parse_name_directive(line, "run-rustfix")
     }
+
+    fn parse_edition(&self, line: &str) -> Option<String> {
+        match line {
+            ed @ "2015" | "2018" => Some(ed.to_owned()),
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/header.rs]: also, maybe `line.trim()`? I don't think we'll have edition names w/ internal whitespace =)

File: src/tools/compiletest/src/runtest.rs [hunk 3]
```diff
@@ -571,6 +571,10 @@ impl<'test> TestCx<'test> {
 
         rustc.args(self.split_maybe_args(&self.config.target_rustcflags));
         rustc.args(&self.props.compile_flags);
+        rustc.arg(&format!(
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/runtest.rs]: I would prefer more like

```rust
if let Some(e) = self.props.editions {
  rustc.arg(&format!("--edition={}", e))
}
```

this way we can test the compiler's default behavior

File: src/tools/compiletest/src/runtest.rs [hunk 4]
```diff
@@ -1370,7 +1374,11 @@ impl<'test> TestCx<'test> {
             .arg("-o")
             .arg(out_dir)
             .arg(&self.testpaths.file)
-            .args(&self.props.compile_flags);
+            .args(&self.props.compile_flags)
+            .arg(&format!(
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/runtest.rs]: (same here)

File: src/tools/compiletest/src/runtest.rs [hunk 5]
```diff
@@ -1765,6 +1773,10 @@ impl<'test> TestCx<'test> {
         }
 
         rustc.args(&self.props.compile_flags);
+        rustc.arg(&format!(
```
Comment (nikomatsakis) [inline src/tools/compiletest/src/runtest.rs]: ...and here. Maybe we want some helper method to make that nicer.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I think it'd be better only to supply the `--edition` argument if `edition:foo` appears in the test. Otherwise, looks great!

Also, I am not sure whether  `-Zunstable-options` are needed or not =)

File: src/tools/compiletest/src/runtest.rs (hunk 5 above)
Comment (mark-i-m) [reply src/tools/compiletest/src/runtest.rs]: I just made this behave as if the user passed `compile-flags: --edition=2018` instead.

**Comment by mark-i-m** [comment]
@nikomatsakis  Ok so...
- I think I addressed all your comments. I ended up changing the implementation to be as if the `compile-flags: --edition=2018` option was passed. I think this is a lot cleaner.
- It turns out that `-Zunstable-options` was not needed for any of those tests, so I removed it.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**